File: 0430-Flatten a Multilevel Doubly Linked List/c++/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head) {
        if(!head) return nullptr;
        Node *solution = head;
        stack<Node*> s;
        while(head) {
            if(head->child) {
                if(head->next)
                    s.push(head->next);
                head->next = head->child;
                head->next->prev = head;
                head->child = nullptr;
            } 
            if(!head->next) break; 
            head = head->next;  
        }
        while(!s.empty()) {
            Node *tmp = s.top(); s.pop();
            head->next = tmp;
            tmp->prev = head;
            while(head->next) head = head->next;
        }
        return solution;
    }
};

// Runtime: 4 ms, faster than 97.71% of C++ online submissions for Flatten a Multilevel Doubly Linked List.
// Memory Usage: 7.6 MB, less than 15.33% of C++ online submissions for Flatten a Multilevel Doubly Linked List.

File: 0430-Flatten a Multilevel Doubly Linked List/golang/solution.go
package golang

type Node struct {
	Val   int
	Prev  *Node
	Next  *Node
	Child *Node
}

func flatten(root *Node) *Node {
	solution := root
	var stack []*Node
	for root != nil {
		if root.Child != nil {
			if root.Next != nil {
				stack = append(stack, root.Next)
			}
			root.Next = root.Child
			root.Next.Prev = root
			root.Child = nil
		}

		if root.Next == nil {
			break
		}
		root = root.Next
	}

	for len(stack) != 0 {
		top := len(stack) - 1
		root.Next = stack[top]
		root.Next.Prev = root
		stack = stack[:top]
		for root.Next != nil {
			root = root.Next
		}
	}
	return solution
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Flatten a Multilevel Doubly Linked List.
// Memory Usage: 2.9 MB, less than 70.27% of Go online submissions for Flatten a Multilevel Doubly Linked List.
